A stepdown might result in inconsistent log in the old leader

**Describe the bug**

During the randomized raft test observed the following inconsistency among the logs :
```
java.lang.AssertionError: Expected to find the same entry at a committed index on all nodes, but found {2-partition-1=Indexed{index=137, entry=ZeebeEntry{term=30, timestamp=1600257494188, lowestPosition=0, highestPosition=1}}, 0-partition-1=Indexed{index=137, entry=ZeebeEntry{term=30, timestamp=1600257494188, lowestPosition=0, highestPosition=1}}, 1-partition-1=Indexed{index=137, entry=InitializeEntry{term=78, timestamp=2020-09-16 01:58:14,425}}}
```
On investigating further, it was observed that the old leader (server 1) appended an entry generated by itself with the new term. So when it receives append requests from the new leader, it immediately accepted the appends with out truncating the entries that it has appended during it's term but not committed.

```
13:58:14.426 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Appended Indexed{index=137, entry=InitializeEntry{term=78, timestamp=2020-09-16 01:58:14,425}}
```
Node 1 Appended more events, but never committed anything. Then later
```
13:58:14.443 [] DEBUG io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Received ConfigureRequest{term=80, leader=2, index=0, timestamp=1600257493772, members=[DefaultRaftMember{id=2, type=ACTIVE, updated=2020-09-16T11:58:13.772246Z}, DefaultRaftMember{id=1, type=ACTIVE, updated=2020-09-16T11:58:13.772246Z}, DefaultRaftMember{id=0, type=ACTIVE, updated=2020-09-16T11:58:13.772246Z}]}
13:58:14.443 [] TRACE io.atomix.raft.storage.system.MetaStore - Store term 80
13:58:14.444 [] TRACE io.atomix.raft.storage.system.MetaStore - Store vote null
13:58:14.444 [] DEBUG io.atomix.raft.impl.RaftContext - RaftServer{1-partition-1} - Set term 80
13:58:14.444 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1-partition-1} - Found leader 2
13:58:14.444 [] TRACE io.atomix.raft.storage.system.MetaStore - Store vote null
13:58:14.444 [] DEBUG io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Sending ConfigureResponse{status=OK}
13:58:14.444 [] INFO  TEST - Running deliverNextMessage to 2 on 2
13:58:14.444 [] INFO  TEST - Running runNextTask on 2
13:58:14.444 [] TRACE io.atomix.raft.roles.LeaderAppender - RaftServer{2-partition-1} - Received ConfigureResponse{status=OK} from 1
13:58:14.444 [] INFO  TEST - Running tickElectionTimeout on 1
13:58:14.444 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Appended Indexed{index=146, entry=ZeebeEntry{term=80, timestamp=1600257494444, lowestPosition=0, highestPosition=1}}
13:58:14.444 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1-partition-1} - Transitioning to FOLLOWER
13:58:14.445 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Cancelling append timer
13:58:14.445 [] DEBUG io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Received AppendRequest{term=80, leader=2, prevLogIndex=142, prevLogTerm=30, entries=0, commitIndex=136}
13:58:14.445 [] DEBUG io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Rejected AppendRequest{term=80, leader=2, prevLogIndex=142, prevLogTerm=30, entries=0, commitIndex=136}: Previous entry term (30) does not match local log's term for the same entry (78)
13:58:14.445 [] DEBUG io.atomix.raft.roles.LeaderRole - RaftServer{1-partition-1}{role=LEADER} - Sending AppendResponse{status=OK, term=80, succeeded=false, lastLogIndex=141, lastSnapshotIndex=0}
13:58:14.445 [] INFO  TEST - Running deliverNextMessage to 2 on 2
13:58:14.445 [] INFO  TEST - Running runNextTask on 2
13:58:14.445 [] TRACE io.atomix.raft.roles.LeaderAppender - RaftServer{2-partition-1} - Received AppendResponse{status=OK, term=80, succeeded=false, lastLogIndex=141, lastSnapshotIndex=0} from 1
13:58:14.445 [] TRACE io.atomix.raft.roles.LeaderAppender - RaftServer{2-partition-1} - Reset next index for RaftMemberContext{member=1, term=80, configIndex=0, snapshotIndex=0, nextSnapshotIndex=0, nextSnapshotChunk=null, matchIndex=0, nextIndex=142, heartbeatTime=1600257494441, appending=0, appendSucceeded=false, appendTime=1600257494441, configuring=false, installing=false, failures=0} to 142
13:58:14.445 [] TRACE io.atomix.raft.roles.LeaderAppender - RaftServer{2-partition-1} - Sending AppendRequest{term=80, leader=2, prevLogIndex=141, prevLogTerm=30, entries=2, commitIndex=136} to 1
```
Node 1 appended a index-146 in the role leader. The entry is generated by node-1 not the one replicated from the new leader. It rejects the following append because the entries doesn't match (expected).

```
13:58:14.449 [] TRACE io.atomix.raft.roles.FollowerRole - RaftServer{1-partition-1}{role=FOLLOWER} - Cancelling heartbeat timer
13:58:14.449 [] DEBUG io.atomix.raft.roles.FollowerRole - RaftServer{1-partition-1}{role=FOLLOWER} - Received AppendRequest{term=80, leader=2, prevLogIndex=143, prevLogTerm=80, entries=0, commitIndex=136}
13:58:14.449 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1-partition-1} - Found leader 2
13:58:14.449 [] TRACE io.atomix.raft.storage.system.MetaStore - Store vote null
13:58:14.450 [] DEBUG io.atomix.raft.roles.FollowerRole - RaftServer{1-partition-1}{role=FOLLOWER} - Rejected AppendRequest{term=80, leader=2, prevLogIndex=143, prevLogTerm=80, entries=0, commitIndex=136}: Previous entry term (80) does not match local log's term for the same entry (78)
13:58:14.450 [] DEBUG io.atomix.raft.roles.FollowerRole - RaftServer{1-partition-1}{role=FOLLOWER} - Sending AppendResponse{status=OK, term=80, succeeded=false, lastLogIndex=142, lastSnapshotIndex=0}

13:58:14.451 [] TRACE io.atomix.raft.roles.FollowerRole - RaftServer{0-partition-1}{role=FOLLOWER} - Committed entries up to index 143
```

Much later, node 1 receives append request from a new leader. Immediately accepts it because the term of previous entry matches even though the entry is wrong.

```
13:58:14.710 [] DEBUG io.atomix.raft.roles.FollowerRole - RaftServer{1-partition-1}{role=FOLLOWER} - Received AppendRequest{term=128, leader=2, prevLogIndex=146, prevLogTerm=80, entries=0, commitIndex=149}
13:58:14.711 [] TRACE io.atomix.raft.roles.FollowerRole - RaftServer{1-partition-1}{role=FOLLOWER} - Committed entries up to index 146
13:58:14.711 [] DEBUG io.atomix.raft.roles.FollowerRole - RaftServer{1-partition-1}{role=FOLLOWER} - Sending AppendResponse{status=OK, term=128, succeeded=true, lastLogIndex=146, lastSnapshotIndex=0}
```

**To Reproduce**

It can be reproduced by running io.atomix.raft.RandomizedRaftTest in branch https://github.com/zeebe-io/zeebe/tree/dd-test-raft

**Expected behavior**

If an index is committed, all nodes should have the same entry at that index.

--- a/atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java
+++ b/atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java
@@ -29,6 +29,8 @@ import io.atomix.raft.cluster.impl.RaftMemberContext;
 import io.atomix.raft.impl.RaftContext;
 import io.atomix.raft.protocol.AppendRequest;
 import io.atomix.raft.protocol.AppendResponse;
+import io.atomix.raft.protocol.ConfigureRequest;
+import io.atomix.raft.protocol.ConfigureResponse;
 import io.atomix.raft.protocol.JoinRequest;
 import io.atomix.raft.protocol.JoinResponse;
 import io.atomix.raft.protocol.LeaveRequest;
@@ -460,6 +462,14 @@ public final class LeaderRole extends ActiveRole implements ZeebeLogAppender {
                         configuring = 0;
                       });
             });
+  }
+
+  @Override
+  public CompletableFuture<ConfigureResponse> onConfigure(final ConfigureRequest request) {
+    if (updateTermAndLeader(request.term(), request.leader())) {
+      raft.transition(Role.FOLLOWER);
+    }
+    return super.onConfigure(request);
   }
 
   @Override

